felix/nftables/maps.go

```diff
@@ -0,0 +1,677 @@
+		return
+	}
+	membersTracker := s.mainSetNameToMembers[mapName]
+	canonMembers.Iter(func(member MapMember) error {
+		s.maybeDecrefChain(member)
+		membersTracker.Desired().Delete(member)
+		return nil
+	})
+	s.updateDirtiness(mapName)
+}
+
+func (s *Maps) GetIPFamily() ipsets.IPFamily {
+	return s.IPVersionConfig.Family
+}
+
+func (s *Maps) filterAndCanonicaliseMembers(mtype MapType, members map[string][]string) set.Set[MapMember] {
+	filtered := set.New[MapMember]()
+	for k, v := range members {
+		filtered.Add(CanonicaliseMapMember(mtype, k, v))
+	}
+	return filtered
+}
+
+// tryResync attempts to bring our state into sync with the dataplane.  It scans the contents of the
+// maps in the dataplane and queues up updates to any maps that are out-of-sync.
+func (s *Maps) LoadDataplaneState() error {
+	// Log the time spent as we exit the function.
+	resyncStart := time.Now()
+	defer func() {
+		s.logCxt.WithFields(logrus.Fields{
+			"resyncDuration":         time.Since(resyncStart),
+			"mapsWithDirtyMembers":   s.mapsWithDirtyMembers.Len(),
+			"mapsToCreateOrRecreate": s.mapNameToProgrammedMetadata.PendingUpdates().Len(),
+			"mapsToDelete":           s.mapNameToProgrammedMetadata.PendingDeletions().Len(),
+		}).Debug("Finished Maps resync")
+	}()
+
+	// Clear the dataplane metadata view, we'll build it back up again as we scan.
+	s.mapNameToProgrammedMetadata.Dataplane().DeleteAll()
+
+	// Load from the dataplane. Update our Dataplane() maps with the actual contents
+	// of the data plane.
+	//
+	// For any map that doesn't match the desired data plane state, we'll queue up an update.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	maps, err := s.nft.List(ctx, "map")
+	if err != nil {
+		if knftables.IsNotFound(err) {
+			// Table doesn't exist - nothing to resync.
+			return nil
+		}
+		return fmt.Errorf("error listing nftables maps: %s", err)
+	}
+
+	// We'll process each map in parallel, so we need a struct to hold the results.
+	// Once knftables is augmented to support reading many maps at once, we can remove this.
+	type mapData struct {
+		name  string
+		elems []*knftables.Element
+		err   error
+	}
+	mapsCh := make(chan mapData)
+	defer close(mapsCh)
+
+	// Start a goroutine to list the elements of each map. Limit concurrent map reads to
+	// avoid spawning too many goroutines if there are a large number of maps.
+	routineLimit := make(chan struct{}, 100)
+	defer close(routineLimit)
+	for _, name := range maps {
+		// Wait for room in the limiting channel.
+		routineLimit <- struct{}{}
+
+		// Start a goroutine to read this map.
+		go func(name string) {
+			// Make sure to indicate that we're done by removing ourselves from the limiter channel.
+			defer func() { <-routineLimit }()
+
+			elems, err := s.nft.ListElements(ctx, "map", name)
+			if err != nil {
+				mapsCh <- mapData{name: name, err: err}
+				return
+			}
+			mapsCh <- mapData{name: name, elems: elems}
```

Comment:
Think this will deadlock if there are >100 maps.  There's nothing reading from this chan until after the loop finishes.

Worth checking the sets code for similar if this is copied from there.